### DISCUSSION TITLE
Added: LDM Notes for November 10th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#5414] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-11-10.md

## Agenda

1. Self types

**Comment by Richiban** [comment]
> several members of the LDM are not convinced by the prevalence of the need to call these implementations once .NET 7 is out

.NET 7? I can't find much info about this... Is there a headline feature we should know about here?

**Reply by DavidArno** [reply]
The .NET road map will see .NET 7, which will have only short term support of 18 months, being released in November 2022. 

https://github.com/dotnet/core/blob/main/roadmap.md

**Reply by Richiban** [reply]
I was aware of the rough timescale for its release; I was asking more about what in .NET 7 is relevant to the topic of self types?

**Reply by MgSam** [reply]
They want to ship the generic math capabilities.

**Comment by huoyaoyuan** [comment]
Is there any opportunity to unify existing interfaces like `IEquatable<T>`? Barrier in ecosystem isn't a good experience.

**Comment by MgSam** [comment]
The choice isn't addressed in the notes directly, but I vastly prefer the `interface IAdditiveIdentity<TSelf> where TSelf : this` form over `interface IAdditionOperators<this TSelf, TOther, TResult>`. There's no reason to introduce a second constraint syntax, especially given how confusing generic definitions are to many developers already.

**Reply by 333fred** [reply]
We never really talked about syntax in the meeting, which is why there isn't any mention in the notes. Personally, I agree.

**Comment by sab39** [comment]
For better or worse, I've spent a lot of years on a codebase that uses the CRTP extensively (along with doing lots of other questionable things with generic types because I tried to be too clever for my own good, but that's another story). One pain point with the CRTP that isn't addressed directly in the notes is that it doesn't allow treating `this` as a `T`, or even casting it particularly easily. I had a bunch of code along the lines of

```cs
public abstract class ThingBase<T>
    where T : ThingBase<T>, IThing<T>
{
    protected T This => (T)(object)this;
}
```

It would have been much more pleasant working in that codebase if I could include `where T : this` to the constraints and treat `this` as a `T` directly.

**Reply by 333fred** [reply]
I don't think this has been among the use-cases brought up so far, it's an interesting one. @tannergooding and @AlekseyTs, making sure you see this.

**Reply by sab39** [reply]
If it would help, I can look through that code and find some of the other weird things I did to workaround not having self types, although off the top of my head most of them overlap with other proposals to some extent (and many of them I'm quite happy to concede were just plain bad ideas in the first place).

I do seem to have a talent for immediately stress-testing the corner cases of every language and type system I use - within an hour of trying C#10 I had 10ish lines of code that reproducibly crash VS22. I'm not sure whether that means I'm a really good resource for language design feedback or a really bad one!

**Reply by 333fred** [reply]
As long as you filled a bug, a good one 😉.

**Reply by sab39** [reply]
[I did](https://developercommunity.visualstudio.com/t/Crash-when-renaming-inner-class-with-inh/1555841), but it was initially closed as needing more info. I followed up once but I didn't notice there'd been more messages since - I guess I need to do something to sign up for notifications emails!

**Reply by bbarry** [reply]
I was thinking about that a little; what would that mean for DIMs:

```csharp
public interface IThing<T>
    where T : IThing<T>
{
    protected T This() => (T)(object)this;
}
```


is valid code today and I believe it means "box `this` and return it as a `T`"

Would the hypothetical:


```csharp
public interface IThing<T>
    where T : this
{
    protected T This() => this;
}
```

need to mean the same thing?

**Reply by sab39** [reply]
That's a great question - is there any way that the box can be avoided if `IThing<T>` is implemented by a struct?
